src/Windows/Host/Client/Impl/BrokerServices/RemotingWebServer.cs

```diff
@@ -1,15 +1,55 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using System.Threading;
-using System.Threading.Tasks;
+using Microsoft.Common.Core.IO;
 using Microsoft.Common.Core.Logging;
+using Microsoft.Common.Core.Services;
 using Microsoft.R.Host.Client.BrokerServices;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace Microsoft.R.Host.Client {
     public class RemotingWebServer : IRemotingWebServer {
-        public Task<string> CreateWebServerAsync(string remoteUrl, string baseAddress, string name, IActionLog log, IConsole console, CancellationToken ct = default(CancellationToken)) {
-            return WebServer.CreateWebServerAsync(remoteUrl, baseAddress, name, log, console, ct);
+        private readonly IServiceContainer _services;
+        private readonly IFileSystem _fs;
+        private readonly IActionLog _log;
+
+        private LocalStaticFileServer _localStaticFileServer;
+        private object _localStaticFileServerLock = new object();
+
+        private RemoteStaticFileServer _remoteStaticFileServer;
+        private object _remoteStaticFileServerLock = new object();
+
+        public RemotingWebServer(IServiceContainer services) {
+            _services = services;
+            _fs = _services.FileSystem();
+            _log = _services.Log();
+        }
+
+        public Task<string> HandleRemoteWebUrlAsync(string remoteUrl, string baseAddress, string name, IConsole console, CancellationToken ct = default(CancellationToken)) {
+            return WebServer.CreateWebServerAndHandleUrlAsync(remoteUrl, baseAddress, name, _log, console, ct);
+        }
+
+        public Task<string> HandleLocalStaticFileUrlAsync(string url, IConsole console, CancellationToken ct = default(CancellationToken)) {
+            CreateLocalStaticFileServer(console);
+            return _localStaticFileServer.HandleUrlAsync(url, ct);
+        }
+
+        public Task<string> HandleRemoteStaticFileUrlAsync(string url,  IRSessionProvider rSessionProvider, IConsole console, CancellationToken ct = default(CancellationToken)) {
+            CreateRemoteStaticFileServerAsync(rSessionProvider, console);
+            return _remoteStaticFileServer.HandleUrlAsync(url, ct);
+        }
+
+        private void CreateLocalStaticFileServer(IConsole console) {
```

Comment:
This should probably be called something like `EnsureLocalStaticFileServer`.

Better yet, make it `Get...` and return the instance (whether created or not). Then the methods above can be one-liners, just using the returned value.